Local disk CAS

ola-rozenfeld opened the pull request:
Implements an optional local disk cache (LRU, CAS) used for remote action outputs.
Goma has this feature, and many of our Reclient customers requested it since (usually because of slow internet connections).

In my benchmarks on Linux with a relatively good internet connection this was not able to reduce the runtime of a fully cached build by a statistically significant amount. But it didn't hurt either.

ola-rozenfeld commented:
Hang on, tests pass on Mac but not on Linux, checking...

ola-rozenfeld commented:
Update: Brave have tried this change and see no difference, because we still need to do an Action Cache RPC. I will now add optional Action Cache to this change to see how much improvement they get with it.

gkousik commented:
Thanks for the PR Ola! I generally like and agree with having a local disk cache feature, but I want to ask some due diligence questions.

1. In slow internet connection scenarios, has racing exec strategy proven to be helpful? Is local disk cache faster than racing in those scenarios? 
2. > In my benchmarks on Linux with a relatively good internet connection this was not able to reduce the runtime of a fully cached build by a statistically significant amount. But it didn't hurt either.

    This is very interesting - what was the size of your disk cache and what was the total output downloaded by the build? I am assuming this is a Chrome build? Its surprising that `network transfer + write` ~= `local_disk_copy` for a full build.
3. I think `StoreCas` also could suffer from a race condition where the local build modifies the file after writing it out. Perhaps we need to validate digest after `cp` is finished?
4. Finally a request, assuming we go ahead (I just want to know answers to (1) and (2)), could this be split up into smaller CLs? I have some readability comments that maybe easier to do with smaller changes :)

ola-rozenfeld commented:
Hey Kousik and Anas! Good questions! 1 - @goodov from Brave has tried this out on a slow connection -- maybe Aleksey can comment on how it compares to racing, because I know he has tried that as well (although it's not yet the default setting we recommend to customers). I'm wondering if there's a way for me to artificially create a VM with a slow connection to try it out -- maybe get a VM in the farthest possible region from our backend?...

2 - I tried a Chromium build on Linux with 50gb cache size, and the actual cache size was 26GB. Which, btw, implies that RBE download metrics that it displays at the end of the build are waaaay off (it usually says `RBE Stats: down 5.78 GB`).

3 - Good point! I didn't think the output files would be locally modified (do they? Do you know which actions do that?). I'll add handling of this case and a unit test. But I'm struggling to imagine this being a race condition unless it's literally two concurrent actions that output the same file -- and if that's the case, isn't it already a race condition in the build itself?...

4 - Yes, agreed, I did it in one go because I wanted @goodov to be able to give it a try easily. I think this can naturally break into 3 PRs -- first, introduce the diskcache with CAS-only plus unit tests, then second PR adds Action Cache functionality plus unit tests, then 3rd PR introduces the client flags plus instrumentation to actually use the diskcache in the client. Does that sound good?

Thank you!

ola-rozenfeld commented:
Updates:
- changed the directory structure following Marc-Antoine's (@maru3l) suggestion (based on Git and Isolate implementations) -- am now splitting the cache directory into 256 subdirectories based on 2-byte prefix of the digest, and using the other bytes for the name. Also moved the "_ac" to suffix. Both of these should improve performance, especially on Windows.
- Split this PR out into https://github.com/bazelbuild/remote-apis-sdks/pull/530 (will send out part 2 after part 1 is in).

I'm leaving this PR here for reference to the whole thing.

ola-rozenfeld commented:
@mrahs Friendly ping! :-) Let me know if you prefer to review the whole thing or in chunks (e.g. https://github.com/bazelbuild/remote-apis-sdks/pull/530).

Thank you!

mrahs commented:
> @mrahs Friendly ping! :-) Let me know if you prefer to review the whole thing or in chunks (e.g. #530).
> 
> Thank you!

Sorry I'm late. I'm trying to understand the context of such cache within past discussions and future requirements.

>I tried a Chromium build on Linux with 50gb cache size, and the actual cache size was 26GB. Which, btw, implies that RBE download metrics that it displays at the end of the build are waaaay off (it usually says RBE Stats: down 5.78 GB).

Ah, interesting. Any chance you've been able to spot the cause of this mismatch?

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+				}
+				it := &qitem{
+					key: k,
+					lat: fileInfoToAccessTime(info),
```
Why use the last access time instead of last modified time?
I don't believe that using the file system to track this information is a good idea. My implementations saved this as a dumb json file and that was very simple, I don't care about the actual time stamp, just the ordering; thus the file is a flat list of files in LRU for eviction order. This also removes the need for the priorityQueue type.

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+				d.mu.Lock()
+				it := heap.Pop(d.queue).(*qitem)
+				d.mu.Unlock()
+				size, err := d.getItemSize(it.key)
```
I don't understand why this does a stat on the file when the size is already in the digest.

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+				diskOpsStart := time.Now()
+				// We only delete the files, and not the prefix directories, because the prefixes are not worth worrying about.
+				if err := os.Remove(d.getPath(it.key)); err != nil {
+					log.Errorf("Error removing file: %w", err)
```
It'd be good if the error would be accumulated, so that Shutdown() could be renamed to Close() and return the accumulated errors.

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+	return key{digest: dg, isCas: !isAc}, nil
+}
+
+func (d *DiskCache) getPath(k key) string {
```
Optional nit: I would have preferred files to have a uniform file name "%s-%d" % (hash, size), instead of having the size as an extension. Then you could use ".cas" as the marker for cas files. IMHO it would be a bit cleaner.

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+	if q.n == cap(q.items) {
+		// Resize the queue
+		old := q.items
+		q.items = make([]*qitem, 2*cap(old)) // Initial capacity needs to be > 0.
```
```
if old == 0 {
  old = 256
}
```
then you can remove the comment

maruel commented on go/pkg/diskcache/diskcache.go:
```diff
@@ -0,0 +1,467 @@
+
+type key struct {
+	digest digest.Digest
+	isCas  bool
```
Is isCas meant to do a performance optimization later? I don't see any behavior change based on this member.

ola-rozenfeld commented:
Thank you, Mark-Antoine -- I'm still tinkering with the whole thing, I just added stats and benchmark tests (not pushed yet), and I'm trying out different variations to see what's fastest. I hope I can get it all done by Monday.

maruel commented on go/pkg/client/cas_download.go:
```diff
@@ -129,6 +140,11 @@ func (c *Client) DownloadOutputs(ctx context.Context, outs map[string]*TreeOutpu
 			return fullStats, err
 		}
+		if c.diskCache != nil {
+			if err := c.diskCache.StoreCas(output.Digest, absPath); err != nil {
```
Do you think it could be done as part of DownloadFiles() ? Otherwise this adds a serial latency.

ola-rozenfeld commented:
I'm really sorry, but I have to admit to myself and others that I apparently have zero spare cycles to put into this in the foreseeable future :-(
Interested parties, such as @maruel or @goodov are welcome to pick up this work and go from here, merging it in any form/parts. Again, I'm sorry! In retrospect, I should have tried to merge and iterate rather than address all comments at once -- but you all had extremely helpful comments, which were important to address.
